teps/0046-triggertemplate-parameter-escaping.md

```diff
@@ -0,0 +1,271 @@
+
+```go
+paramValue := strings.Replace(param.Value, `"`, `\"`, -1)
+```
+
+Unfortunately this replacement was made without context, which lead to issues
+like [#777 "Backslash quote in json will crash the trigger parser"](https://github.com/tektoncd/triggers/issues/777)
+
+In version 0.10.0 this feature was made optional, with users who wanted to
+retain the behaviour requiring to add an annotation to the TriggerTemplate to
+indicate that it should retain "the old escaping behaviour", this was seen as a
+stop-gap until we could understand better the requirements for escaping.
+
+A CEL function was added that allowed marshaling a JSON object as a string,
+which is somewhat useful, but is a barrier to entry for users who just want to
+pass JSON documents through to tasks securely.
+
+### Goals
+
+ * Make it easy to safely pass incoming JSON documents received by
+   EventListeners as parameters to PipelineRuns and TaskRuns, by means of
+   TriggerTemplates.
+
+### Non-Goals
+
+  * Providing for multiple forms of escaping data passing through
+    TriggerTemplates.
+
+### Use Cases (optional)
+ 
+## Requirements
+
+The basic requirement is that a JSON document including escaped quotes within a
+value should be able to rendered into a TriggerTemplate without generating
+invalid JSON.
+
+## Proposal
+
+This proposal seeks to add a field to TriggerTemplate ParamSpec structs, to
+allow a user to mark the field as "requiring to be escaped".
+
+At the point where the parameter's value is being inserted into the
+TriggerTemplate resourceTemplate, if the parameter is flagged for escaping, it
+should be inserted in such a way that the string won't break the subsequent
+parsing of the rendered template.
+
+### Notes/Caveats (optional)
+
+Ultimately, this is about escaping external data being received into the system,
+it does feel somewhat odd that the template insertion has to know whether or not
+the data should be escaped, clearly you could use a parameter in multiple places
+within a document, and in some of those places, escaping might be required or
+more appropriate than others, but it would be configured at the TriggerTemplate
+level, and not at the insertion level.
+
+### Risks and Mitigations
+
+We already have escaping, which can now be disabled, and the proposal is really
+just a fine-tuning of the functionality to allow users to control which strings
+they need to escape.
+
+### User Experience (optional)
+
+### Performance (optional)
+
+There's unlikely to be a significant performance hit from this approach, as the
+functionality already exists.
+
+## Design Details
+
+This would extend the [current](https://github.com/tektoncd/triggers/blob/master/pkg/apis/triggers/v1alpha1/param.go#L5-L15) `ParamSpec` struct.
+
+```
+type ParamSpec struct {
+    < existing fields omitted >>
+
+	// Escape indicates whether or not the value should be escaped as JSON
+	// before inserting into the templated bodies.
+	Escape bool `json:"escape,omitempty"`
```

Comment:
Escape is a bit  generic...what kind of escaping does it mean? I think the alternative where we have different  "strategies" sounds a bit more descriptive.